server/src/main/java/io/spine/server/bus/Bus.java

```diff
@@ -99,4 +204,127 @@ public void unregister(D dispatcher) {
+     * @return the message itself if it passes the filtering or
+     * {@link Optional#absent() Optional.absent()} otherwise
+     */
+    private Iterable<T> filter(Iterable<T> messages, StreamObserver<IsSent> observer) {
+        checkNotNull(messages);
+        checkNotNull(observer);
+        final Collection<T> result = new LinkedList<>();
+        for (T message : messages) {
+            final Optional<IsSent> response = preProcess(toEnvelope(message));
+            if (response.isPresent()) {
+                observer.onNext(response.get());
+            } else {
+                result.add(message);
+            }
+        }
+        return result;
+    }
+
+    /**
+     * Pre-processes the given message.
+     *
+     * <p>The pre-processing may include validation.
+     *
+     * <p>If the given message is completely processed and should not be passed to the dispatchers
+     * via {@link #doPost doPost} method, the returned {@link Optional} contains a value with either
+     * status.
+     *
+     * <p>If the message should be passed to the dispatchers via {@link #doPost doPost}, the result
+     * of this method is {@link Optional#absent() Optional.absent()}.
+     *
+     * @param message the {@linkplain MessageEnvelope message envelope} to pre-process
+     * @return the result of message processing by this bus if any, or
+     * {@link Optional#absent() Optional.absent()} otherwise
+     */
+    protected abstract Optional<IsSent> preProcess(E message);
+
+    /**
+     * Packs the given message of type {@code T} into an envelope of type {@code E}.
+     *
+     * @param message the message to pack
+     * @return new envelope with the given message inside
+     */
+    protected abstract E toEnvelope(T message);
+
+    /**
+     * Posts the given envelope to the bus.
+     *
+     * <p>Finds and invokes the {@linkplain MessageDispatcher MessageDispatcher(s)} for the given
+     * message.
+     *
+     * <p>This method assumes that the given message has passed the filtering.
+     *
+     * @return the result of mailing with the Message ID and:
+     *         <ul>
+     *             <li>{@link io.spine.base.Status.StatusCase#OK OK} status if the message has been
+     *                 passed to the dispatcher;
+     *             <li>{@link io.spine.base.Failure Failure} status, if a {@code Failure} has
+     *                 happened during the message handling (if applicable);
+     *             <li>{@link io.spine.base.Error Error} status if a {@link Throwable}, which is not
+     *                 a {@link io.spine.base.FailureThrowable FailureThrowable}, has been thrown
+     *                 during the message posting.
+     *         </ul>
+     * @see #post(Message, StreamObserver) for the public API
+     */
+    protected abstract IsSent doPost(E envelope);
+
+    /**
+     * Posts each of the given envelopes into the bus and acknowledges the message posting with
```

Comment:
`Posts each of the given envelopes into the bus and notifies the given observer.`